File: Data Structures & Algorithms/maximum-sum-circular-subarray/submission-3.go
func maxSubarraySumCircular(nums []int) int {
	// Edge case for empty array
    if len(nums) == 0{
        return 0
    }

    // Init global max/min with the 1st element
    globalMax, globalMin := nums[0], nums[0]
    curMax, curMin := 0, 0
    total := 0

    for _, n := range nums{
        // Kadane's logic for Max and Min
        curMax = max(curMax + n, n)
        curMin = min(curMin + n, n)

        total += n

        globalMax = max(globalMax, curMax)
        globalMin = min(globalMin, curMin)
    }

    // If all numbers are negative, globalMax will be max single element.
    // Otherwise, we compare the standard max vs. the circular max (total - globalMin)
    if globalMax > 0 {
        return max(globalMax, total - globalMin)
    }
    return globalMax
}

File: Data Structures & Algorithms/sliding-window-maximum/submission-3.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        res = []
        l = 0
        curMax = float("-inf")
        for r in range(len(nums)):
            curMax = max(curMax, nums[r])
            if r - l + 1 < k:
                continue
            res.append(curMax)
            if curMax == nums[l]:
                curMax = float("-inf")
                for i in range(l + 1, r + 1):
                    curMax = max(curMax, nums[i])
            l += 1
        return res